python/ray/tune/suggest/ax.py

```diff
@@ -0,0 +1,113 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+try:
+    import ax.service.ax_client as ax_client
+except ImportError:
+    ax_client = None
+
+from ray.tune.suggest.suggestion import SuggestionAlgorithm
+
+
+class AxSearch(SuggestionAlgorithm):
+    """A wrapper around Ax to provide trial suggestions.
+
+    Requires Ax to be installed.
+    Ax is an open source tool from Facebook for configuring and
+    optimizing experiments. More information can be found in https://ax.dev/.
+
+    Parameters:
+        parameters (list[dict]): Parameters in the experiment search space.
+            Required elements in the dictionaries are: "name" (name of
+            this parameter, string), "type" (type of the parameter: "range",
+            "fixed", or "choice", string), "bounds" for range parameters
+            (list of two values, lower bound first), "values" for choice
+            parameters (list of values), and "value" for fixed parameters
+            (single value).
+        objective_name (str): Name of the metric used as objective in this
+            experiment. This metric must be present in `raw_data` argument
+            to `log_data`.
+        max_concurrent (int): Number of maximum concurrent trials. Defaults
+            to 10.
+        minimize (bool): Whether this experiment represents a minimization
+            problem.
+        parameter_constraints (list[str]): Parameter constraints, such as
+            "x3 >= x4" or "x3 + x4 >= 2".
+        outcome_constraints (list[str]): Outcome constraints of form
+            "metric_name >= bound", like "m1 <= 3."
+        outcome_names (list[str]): Names of outcome constraints, must
+            equal number of outcome_constraints
+
+    Example:
+        >>> parameters = [
+        >>>     {"name": "x1", "type": "range", "bounds": [0.0, 1.0]},
+        >>>     {"name": "x2", "type": "range", "bounds": [0.0, 1.0]},
+        >>> ]
+        >>> algo = AxSearch(parameters=parameters,
+        >>>     objective_name="hartmann6", max_concurrent=4)
+    """
+
+    def __init__(self,
+                 parameters,
+                 objective_name,
+                 max_concurrent=10,
+                 minimize=False,
+                 parameter_constraints=[],
```

Comment:
we should change this to `parameter_constraints=None` (along with any default kwarg that is `[]` or `{}`. There's a weird behavior with these lists being mutable and resulting in some hidden side effects if there are multiple instantiations.

For these defaults, the assignment should be like:
```
self._ax.create_experiment(
            name="ax",
            parameters=parameters,
            objective_name=objective_name,
            minimize=minimize,
            parameter_constraints=parameter_constraints or [],
            outcome_constraints=outcome_constraints or [],
...)
```